### PR TITLE
feat: add notification payment failed

### PR DESCRIPTION
**Pull request by mirado447**

### CODE REVIEW

**Review by haritianaadriano — CHANGES_REQUESTED** [review]
So first part, you cannot re-use the PaidFeeByMpbsNotoficationBody because that one is consumed by the success service mailer. Then you consume it in the failed part it will send too the mail for the success, because the success consumes too the PaidFeeByMpbsNotification

File: src/main/java/school/hei/haapi/service/PaymentService.java [hunk 1]
```diff
@@ -88,8 +89,28 @@ List<Payment> getByStudentIdAndFeeId(String studentId, String feeId) {
 
   public void computeRemainingAmount(String feeId, int amount) {
     Fee associatedFee = feeService.getById(feeId);
+    User student = associatedFee.getStudent();
+    User.Status[] statuses = new User.Status[2];
+    statuses[0] = student.getStatus();
```
Comment (haritianaadriano) [inline src/main/java/school/hei/haapi/service/PaymentService.java]: make some comments for dev readability in your code logic,

File: src/main/java/school/hei/haapi/service/PaymentService.java [hunk 2]
```diff
@@ -88,8 +89,28 @@ List<Payment> getByStudentIdAndFeeId(String studentId, String feeId) {
 
   public void computeRemainingAmount(String feeId, int amount) {
     Fee associatedFee = feeService.getById(feeId);
+    User student = associatedFee.getStudent();
+    User.Status[] statuses = new User.Status[2];
```
Comment (haritianaadriano) [inline src/main/java/school/hei/haapi/service/PaymentService.java]: i think studentStatusBetweenPayingFee is better to mark that it's an array of state of student status

File: src/main/java/school/hei/haapi/service/PaymentService.java [hunk 3]
```diff
@@ -88,8 +89,28 @@ List<Payment> getByStudentIdAndFeeId(String studentId, String feeId) {
...
+    statuses[0] = student.getStatus();
+
     associatedFee.setRemainingAmount(associatedFee.getRemainingAmount() - amount);
-    computeUserStatusAfterPayingFee(associatedFee.getStudent());
+    computeUserStatusAfterPayingFee(student);
+    statuses[1] = student.getStatus();
+    if (SUSPENDED.equals(statuses[0]) && ENABLED.equals(statuses[1])) {
+      Payment payment =
+          Payment.builder()
+              // TIPS: no need  .type()
```
Comment (haritianaadriano) [inline src/main/java/school/hei/haapi/service/PaymentService.java]: say why it's not necessary to not import the type()

File: src/main/java/school/hei/haapi/service/PaymentService.java [hunk 4]
```diff
@@ -88,8 +89,28 @@ List<Payment> getByStudentIdAndFeeId(String studentId, String feeId) {
...
+              .fee(associatedFee)
+              .amount(amount)
+              .creationDatetime(Instant.now())
+              .comment(associatedFee.getComment())
+              .build();
+      SuspensionEndedEmailBody suspensionEndedEmailBody = SuspensionEndedEmailBody.from(payment);
+      eventProducer.accept(List.of(suspensionEndedEmailBody));
+      log.info(
+          "End of suspension notification for user {} sent to Queue.",
+          suspensionEndedEmailBody.getMpbsAuthorEmail());
```
Comment (haritianaadriano) [inline src/main/java/school/hei/haapi/service/PaymentService.java]: make this logic for sending mail in a separate method as notifyStudentForEnabling or something like that

File: src/main/java/school/hei/haapi/service/PaymentService.java [hunk 5]
```diff
@@ -88,8 +89,28 @@ List<Payment> getByStudentIdAndFeeId(String studentId, String feeId) {
...
   public void computeRemainingAmount(String feeId, int amount) {
     Fee associatedFee = feeService.getById(feeId);
+    User student = associatedFee.getStudent();
+    User.Status[] statuses = new User.Status[2];
+    statuses[0] = student.getStatus();
+
     associatedFee.setRemainingAmount(associatedFee.getRemainingAmount() - amount);
-    computeUserStatusAfterPayingFee(associatedFee.getStudent());
+    computeUserStatusAfterPayingFee(student);
+    statuses[1] = student.getStatus();
```
Comment (haritianaadriano) [inline src/main/java/school/hei/haapi/service/PaymentService.java]: log here to and say "User status is computed to ..."

File: src/main/java/school/hei/haapi/service/MpbsVerificationService.java [hunk 6]
```diff
@@ -133,6 +139,21 @@ private MpbsStatus defineMpbsStatusFromOrangeTransactionDetails(
...
+              .amount(mpbs.getAmount())
+              .creationDatetime(Instant.now())
+              .comment(correspondingFee.getComment())
+              .build();
+      PaidFeeByMpbsFailedNotificationBody notificationBody =
+          PaidFeeByMpbsFailedNotificationBody.from(paymentFromMpbs);
+      eventProducer.accept(List.of(notificationBody));
+      log.info(
+          "Failed payment notification for user {} sent to Queue.",
+          notificationBody.getMpbsAuthorEmail());
```
Comment (haritianaadriano) [inline src/main/java/school/hei/haapi/service/MpbsVerificationService.java]: separate this too in a different method

File: src/main/java/school/hei/haapi/service/PaymentService.java [hunk 7]
```diff
@@ -88,8 +89,28 @@ List<Payment> getByStudentIdAndFeeId(String studentId, String feeId) {
 
   public void computeRemainingAmount(String feeId, int amount) {
     Fee associatedFee = feeService.getById(feeId);
+    User student = associatedFee.getStudent();
+    User.Status[] statuses = new User.Status[2];
+    statuses[0] = student.getStatus();
+
     associatedFee.setRemainingAmount(associatedFee.getRemainingAmount() - amount);
-    computeUserStatusAfterPayingFee(associatedFee.getStudent());
+    computeUserStatusAfterPayingFee(student);
```
Comment (haritianaadriano) [inline src/main/java/school/hei/haapi/service/PaymentService.java:105]: create a test for this method: computeUserStatusAfterPayingFee

**Review by haritianaadriano — CHANGES_REQUESTED** [review]

File: src/main/java/school/hei/haapi/service/LetterService.java [hunk 8]
```diff
@@ -126,7 +126,8 @@ public String getBucketKey(String studentRef, String filename) {
   }
 
   public static String generateRef(String id) {
-    return id.substring(0, 6)
+    return "HEI-"
+        + id.substring(0, 6)
         + "-"
```
Comment (haritianaadriano) [inline src/main/java/school/hei/haapi/service/LetterService.java]: Please rebase always your branch, in the case that you tell me that your branch is correctly rebased i'll be ok. But here you didn't touch this file right ?

File: src/test/java/school/hei/haapi/unit/PaymentServiceUnitTest.java [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+    User userWithUnpaidFees = student2();
+    User userWithoutUnpaidFees = student3();
+
+    paymentService.computeUserStatusAfterPayingFee(userWithUnpaidFees);
+    paymentService.computeUserStatusAfterPayingFee(userWithoutUnpaidFees);
+    User updatedUserWithUnpaidFees = userService.findById(userWithUnpaidFees.getId());
+    User updatedUserWithoutUnpaidFees = userService.findById(userWithoutUnpaidFees.getId());
+
+    assertEquals(SUSPENDED, updatedUserWithUnpaidFees.getStatus());
+    assertEquals(ENABLED, updatedUserWithoutUnpaidFees.getStatus());
```
Comment (haritianaadriano) [inline src/test/java/school/hei/haapi/unit/PaymentServiceUnitTest.java:173]: Your test doesn't cover all case. Add the case where the userWithUnpaidFees pay his fees and must be ENABLED

**Review by haritianaadriano — CHANGES_REQUESTED** [review]

File: src/main/java/school/hei/haapi/service/LetterService.java (hunk 8 above)
Comment (mirado447) [reply src/main/java/school/hei/haapi/service/LetterService.java]: actually it was Ilo who asked me to make this modification

**Review by haritianaadriano — APPROVED** [review]

**Pull request merged**